cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/dataHandler.ts

```diff
@@ -0,0 +1,390 @@
+// and calculates the path based on the parentId received
+function findPath(path, parentId, data) {
+  data.forEach((temp) => {
+    if (temp.id === parentId) {
+      path.push(temp.name);
+      if (temp.parentId != null) {
+        findPath(path, temp.parentId, data);
+      }
+    }
+  });
+  return path;
+}
+
+// The function receives an array of flat rows and adds the path property to every row
+export function addPathToFlatMap(data) {
+  data.forEach((tmp) => {
+    if (!tmp.path) {
+      if (tmp.parentId != null) {
+        if (!tmp.path) {
+          tmp.path = [tmp.name];
+        }
+        tmp.path = findPath(tmp.path, tmp.parentId, data).reverse();
+      } else {
+        tmp.path = [tmp.name];
+      }
+    }
+  });
+  return data;
+}
+
+// The function receives a flat schema and is meant to convert it into the
+// correct JSON format that the backend is expecting
+export function toJsonHandler(records) {
+  const data = [...records];
+  const json = {};
+  data.forEach((d) => {
+    if (d.parentId === null) {
+      json[d.name] = {};
+      json[d.name] = jsonRec(d, data, json[d.name]);
+    }
+  });
+  return json;
+}
+
+// A recursive function that receives a parent element, a flat schema and the property in
+// which is meant to add children or assign a value depending on its type
+function jsonRec(d, data, json) {
+  data.map((k) => {
+    if (k.parentId === d.id) {
+      json[k.name] = {};
+      if (k.type === 'record' && !k.path) {
+        jsonRec(k, data, json[k.name]);
+        return;
+      }
+      json[k.name] = k.path;
+    }
+  });
+  return json;
+}
+
+// The function receives json format and is meant to convert it into the
+// correct flat schema format that the frontend will use to render the records
+export function jsonToFlat(json) {
+  let result = [];
+  const keys = Object.keys(json);
+  keys.map((key) => {
+    const uniqueId = uuidV4();
+
+    if (typeof json[key] === 'object') {
+      result.push({ id: uniqueId, name: key, parentId: null, type: 'record' });
+      result = iterateChildren(uniqueId, json[key], result);
+    }
+  });
+  return result;
+}
+
+// This function receives a parentId, and object and an array,
+// it iterates through the keys of the objects and returns a flat row
+function iterateChildren(parentId, tree, result) {
+  const keys = Object.keys(tree);
+
+  keys.map((key: string) => {
+    const uniqueId = uuidV4();
+    if (!Array.isArray(tree[key])) {
+      result.push({ id: uniqueId, name: key, parentId, type: 'record' });
+      iterateChildren(uniqueId, tree[key], result);
+      return;
+    }
+    result.push({ id: uniqueId, name: key, parentId, type: 'string', path: tree[key] });
+  });
+  return result;
+}
+
+// This function receives two arrays, one to push and one to loop, a unique id and a margin number,
+// it returns an array with flat rows
+export const handleChildren = (childrenOfSelected, item, uniqueId, m) => {
```

Comment:
nit: is this function getting children of a field as flat rows? If so please rename the function to `getChildren`

Reply:
Also why does it receive a margin as input? Shouldn't that be a logic specific to rendering? Can we calculate the margin based on the number of parents a particular field has?